File: 041_first_missing_positive/Solution.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        int len = nums.length;

        for (int i = 0; i < len; i++) {
            while (nums[i] > 0 && nums[i] <= len && nums[i] != nums[nums[i] - 1]) {
                swap(nums[i] - 1, i, nums);
            }
        }

        for (int i = 0; i < len; i++) {
            if (nums[i] != i + 1) return i + 1;
        }

        return len + 1;
    }

    public void swap(int i, int j, int[] nums) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{3, 4, -1, 1};
        Solution solution = new Solution();
        System.out.println(solution.firstMissingPositive(nums));
    }
}

File: 041_first_missing_positive/Solution.py
class Solution:
    def firstMissingPositive(self, nums: list) -> int:
        arr_len = len(nums)

        def swap(i, j, nums):
            tmp = nums[i]
            nums[i] = nums[j]
            nums[j] = tmp

        for i in range(arr_len):
            while 0 < nums[i] <= arr_len and nums[i] != nums[nums[i] - 1]:
                swap(i, nums[i] - 1, nums)

        for i in range(arr_len):
            if nums[i] != i + 1: return i + 1

        return arr_len + 1


if __name__ == '__main__':
    nums = [3, 4, -1, 1]
    solution = Solution()
    print(solution.firstMissingPositive(nums))

